File: apps/mobile_flutter/lib/features/profiles/domain/profile_repository.dart
import '../../../core/api/api_client.dart';
import 'learner_profile.dart';

class ProfileRepository {
  const ProfileRepository(this.apiClient);

  final ApiClient apiClient;

  Future<List<LearnerProfile>> fetchProfiles() async {
    final rows = await apiClient.getList('/profiles');
    return rows
        .cast<Map<String, dynamic>>()
        .map(LearnerProfile.fromJson)
        .toList(growable: false);
  }

  Future<LearnerProfile> saveProfile(LearnerProfile profile) async {
    final json = await apiClient.postMap('/profiles', profile.toJson());
    return LearnerProfile.fromJson(json);
  }
}

File: apps/web_pwa/src/features/profiles/profileRepository.ts
import type { ApiClient } from '../../core/api/apiClient';
import { profileFromJson, profileToJson, type LearnerProfile } from './types';

// Mirrors apps/mobile_flutter/lib/features/profiles/domain/profile_repository.dart.
export class ProfileRepository {
  private readonly apiClient: ApiClient;

  constructor(apiClient: ApiClient) {
    this.apiClient = apiClient;
  }

  async fetchProfiles(): Promise<LearnerProfile[]> {
    const rows = await this.apiClient.getList<Record<string, unknown>>('/profiles');
    return rows.map(profileFromJson);
  }

  async saveProfile(profile: LearnerProfile): Promise<LearnerProfile> {
    const json = await this.apiClient.postMap<Record<string, unknown>>(
      '/profiles',
      profileToJson(profile),
    );
    return profileFromJson(json);
  }
}
